### PR TITLE
Added Appveyor support

### PR DESCRIPTION
**Pull request by emoon**
Support for Appveyor (building with Rust on Windows) I can't test this as it needs to be enabled on Appveyor for your repro (and you need to change the `_correct_hash_here` in the commit for it too look correct because I don't know what that is now.

Hopefully this will be a good start :)

### CODE REVIEW

**Comment by yupferris** [comment]
I'm pretty sure Travis also supports windows builds for Rust?

**Comment by emoon** [comment]
Nope, Travis is Mac and Linux.

**Comment by yupferris** [comment]
Man, I could've SWORN I had it running a Windows build agent at some point. I need to look more into this. My mind is literally blown rn haha.

**Comment by emoon** [comment]
https://docs.travis-ci.com/user/multi-os/ (lists linux, osx)

**Comment by yupferris** [comment]
just tried it, and it _does_ fire up workers if you list additional os', but if you read the logs they're also just linux workers: https://travis-ci.org/yupferris/travis-windows-test

This is still super strange to me; I could've SWORN like 2 months ago I had Windows agents running. I remember pretty clearly seeing the Windows icon (and I also see the os x/linux ones got a facelift). Strange.

Anyways, I'll give this a shot then :)

**Comment by emoon** [comment]
Alright :)

**Comment by yupferris** [comment]
hmm wait, is that commit hash going to have to be updated manually all the time? Can't we point to a branch?

**Comment by ryanhiebert** [comment]
I was thinking the same thing at first. I don't think it's a commit hash, I think it's just a uuid for the project.

**Comment by emoon** [comment]
No, that is not a commit hash, you find the link on the Appveyor setup page for the project

**Comment by yupferris** [comment]
Gotcha. Do I have push access on your fork, so that I can update the hash?

**Comment by emoon** [comment]
lets see

**Comment by CryZe** [comment]
Can't you just modify his branch before merging?

Am Mittwoch, 27. Januar 2016 schrieb ferris :

**Comment by emoon** [comment]
Sure, give me the uuid and I will fix

**Comment by yupferris** [comment]
that works too, but I'm still curious about fork permissions in general :) owjloq84v91147nd

**Comment by ryanhiebert** [comment]
I assume that by default you couldn't, because AFAIK there's no branch-level permissions. If that's not the case, that's something I'd be very interested to know as well.

**Comment by CryZe** [comment]
You can just fetch Pull Requests via git, add commits and then merge. Git
allows you to do that.

Am Mittwoch, 27. Januar 2016 schrieb ferris :

**Comment by ryanhiebert** [comment]
@yupferris : I notice that the badges are back below the title in the readme. Did you still want them on the title line?

**Comment by yupferris** [comment]
@CryZe I'm still not sure I understand. Can you give a more detailed example?

@ryanhiebert yeah I was going to bring that up as well :) @emoon I'd like both badges to be on the same line in the readme

**Comment by ryanhiebert** [comment]
@yupferris : @CryZe is just pointing out that you could merge it locally to add minor adjustments like this, rather than using GitHub's merge button.

**Comment by emoon** [comment]
I don't get it. I have it like this in minifb:

```
[![Build Status](https://travis-ci.org/emoon/rust_minifb.svg)](https://travis-ci.org/emoon/rust_minifb)
[![Build Status](https://ci.appveyor.com/api/projects/status/sfvgqq4d4sjulkbx?svg=true)](https://ci.appveyor.com/project/emoon/rust-minifb)
```

and it's on the same line

**Comment by ryanhiebert** [comment]
@emoon : GitHub flavored markdown preserves all line breaks. Also, I think he's wanting them actually _on_ the title line, not below it.

**Comment by yupferris** [comment]
^ yep that

**Comment by ryanhiebert** [comment]
Though, based on your comment, that may have been what you were expecting...

**Comment by yupferris** [comment]
I'm expecting: https://github.com/yupferris/butterball <- badges on the same line as the title

from what I can see in the commit, there's the title on one line, and the batches on the next line.

**Comment by CryZe** [comment]
@yupferris You basically just add his clone URL to your repo as a remote,
then fetch from there, checkout his branch and then commit to that. Then
you can merge it, push it and GitHub will understand what was going on and
resolve this PR properly.

Am Mittwoch, 27. Januar 2016 schrieb Ryan Hiebert :

**Comment by emoon** [comment]
When I access `https://ci.appveyor.com/api/projects/status/owjloq84v91147nd?svg=true` I get 
`{"message":"Build not found or access denied."}`

Some rights that has been missed?

**Comment by yupferris** [comment]
@CryZe so you're effectively adding a branch on your repo that matches names with the branch he's making the PR on on his repo, and the PR understands this?

Also, you say "merge it, push it"; what are you merging here?
- So, I'm on project 'a' on branch 'master'.
- Someone forks 'a' (we'll call this 'a-fork') and creates branch 'a-fork-branch', then creates a PR from this.
- I add 'a-fork' as a remote on my local repo.
- I fetch from 'a-fork' and checkout 'a-fork-branch'.
- I make the necessary changes and commit to my local 'a-fork-branch'.
- (this is where my confusion starts) I push to a new branch 'a-fork-branch' on my repo 'a', so now the latest changes are on 'a'/'a-fork-branch', and these are included in the PR? I'm confused about this step.

**Comment by yupferris** [comment]
Sorry if I'm being thick, the whole thing sounds like magic to me and I want to know exactly what you're describing here :)

**Comment by ryanhiebert** [comment]
They wouldn't be included in the original PR, but when the new branch was merged, the original PR would be closed. That leaves a gap in the testing, which could be remedies by opening a second pull request. As long as the new pull request was not rebased, both the new and old pull requests would be closed when the new pull request was merged, because github will notice when the commits from a pull request are merged, and close it automatically.

**Comment by yupferris** [comment]
@emoon I'm getting the same thing. I'm new to Appveyor; might you know where I can find these permissions?

**Comment by emoon** [comment]
did you take the id from the "sample markdown" line?

**Comment by yupferris** [comment]
I took it from the "branch svg" line, but looks like the uuid is the same in both.

Full "sample markdown" line:
`[Build status](https://ci.appveyor.com/project/yupferris/rustendo64/branch/master)`

**Comment by yupferris** [comment]
I'm also getting a broken image at the top of the "Badges" menu though, perhaps that could be related

**Comment by ryanhiebert** [comment]
Unless I'm reading the config incorrectly, ISTM that it's only testing against Rust nightly. Am I incorrect, or is that intended?

Also, perhaps the master branch has to be built before that icon will resolve.

**Comment by yupferris** [comment]
Yes, it looks like that to me as well. We should rather test on stable. But man, guys, slow down :)

**Comment by yupferris** [comment]
One thing at a time haha

**Comment by ryanhiebert** [comment]
lol

**Comment by emoon** [comment]
Yeah we can change it to stable if wanted. Dunno about the badge though :/ (and yes, it's broken in the readme because it likely also gets no access)

**Comment by emoon** [comment]
Well.. I leave it up to you if you want to merge it in or not and see if it works while building on stable, can otherwise just remove the line from the readme while it's being sorted.

**Comment by yupferris** [comment]
let's change it to stable and I'll merge with the badge in-place, then we'll fix it if that doesn't work

**Comment by yupferris** [comment]
These things tend to be a bit finicky

File: appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+environment:
+  matrix:
+  - TARGET: x86_64-pc-windows-msvc
+  - TARGET: i686-pc-windows-msvc
+  - TARGET: i686-pc-windows-gnu
+install:
+  - ps: Start-FileDownload "https://static.rust-lang.org/dist/rust-1.6.0-${env:TARGET}.exe"
+  - rust-nightly-%TARGET%.exe /VERYSILENT /NORESTART /DIR="C:\Program Files (x86)\Rust"
```
Comment (yupferris) [inline appveyor.yml]: looks like this was the culprit for the failing build: `- rust-nightly-...`

Comment (emoon) [inline appveyor.yml]: Doh! I will have to fix that tomorrow then or you can merge and fix it after the fact
On Jan 27, 2016 23:59, ferris notifications@github.com wrote:In appveyor.yml:

> @@ -0,0 +1,16 @@
> +environment:
> -  matrix:
> -  - TARGET: x86_64-pc-windows-msvc
> -  - TARGET: i686-pc-windows-msvc
> -  - TARGET: i686-pc-windows-gnu
>   +install:
> -  - ps: Start-FileDownload "https://static.rust-lang.org/dist/rust-1.6.0-${env:TARGET}.exe"
> -  - rust-nightly-%TARGET%.exe /VERYSILENT /NORESTART /DIR="C:\Program Files (x86)\Rust"

looks like this was the culprit for the failing build

**Comment by yupferris** [comment]
I'll just wait till you fix it tomorrow :)

**Comment by ryanhiebert** [comment]
On my own fork I set up appveyor, and have been working through these issues, and I've come up with something I'm happy with, and it's based on this work.

https://github.com/ryanhiebert/rustendo64/pull/2
- The build is working
- Only the master branch, and pull requests for it, are built
- The badge in the readme tracks the status of the master branch

I also have a couple of settings enabled, that I highly encourage here:
- Skip branches without `appveyor.yml`
- Rolling builds

The rolling builds is a nice feature that cancels builds that are out-of-date, such as builds for pull requests that have been updated, and quick succession builds on the master branch.

---

There's two ways that we could handle pulling in that work. The first option would be for me to open a pull request to @emoon's branch. That would avoid opening another pull request, but it would require him to merge it into this pull request, which he may not be available to do immediately.

The second would be to open a new pull request on this repository directly from my branch. This would create an extra pull request, but it would show as closing _both_ of the pull requests when it gets merged.

@yupferris : Have a preference?

**Comment by ryanhiebert** [comment]
Or, if you're available to do it, @emoon, you could take my commits from that branch on my repo and add it to your PR. It depends on your availability and preferences.

**Comment by emoon** [comment]
@ryanhiebert done, thanks!

**Comment by emoon** [comment]
Sorry for this somewhat messy PR :)

**Comment by ryanhiebert** [comment]
Eh, that's real life!

**Comment by yupferris** [comment]
Thanks guys! No worries about the messy PR, as long as the code coming in isn't messy ;D

**Comment by yupferris** [comment]
Just looked it over, this is ok to merge, correct? I'd like to do that ASAP.

**Comment by emoon** [comment]
Should be all ready to go now :)

**Comment by ryanhiebert** [comment]
:+1:

**Comment by ryanhiebert** [comment]
@yupferris : Were you able to toggle those two settings on appveyor that I mentioned?

**Comment by ryanhiebert** [comment]
- Rolling builds
- Skip branches without `appveyor.yml`

**Comment by yupferris** [comment]
just enabled them now, that means https://github.com/yupferris/rustendo64/pull/11 won't be checked moving forward unless he rebases, correct?

**Comment by yupferris** [comment]
and same with https://github.com/yupferris/rustendo64/pull/15 ?

**Comment by ryanhiebert** [comment]
Yes, and also that the builds won't fail because they aren't set up correctly.

**Comment by emoon** [comment]
if those PR would re-merge from master it would work fine also

**Comment by yupferris** [comment]
Cool, thanks @ryanhiebert! I'll suggest that @emoon

**Pull request merged**